docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        // functionality (such as go-to-definition).
+        node?: Node;
+    }
+
+    // Declaration information for a symbol, which includes
+    // a list of declarations and potentially synthesized types for those declarations.
+    export interface SymbolDeclInfo {
+        decls: Declaration[];
+        synthesizedTypes: SynthesizedTypeInfo[];
+    }
+
+    // Describes an import
+    export interface ImportedModuleDescriptor {
+        leadingDots: number;
+        nameParts: string[];
+    }
+
+    // Options for resolving an alias declaration.
+    export interface ResolveAliasOptions {
+        resolveLocalNames?: boolean; // Whether to resolve local names in the alias declaration.
+        allowExternallyHiddenAccess?: boolean; // Whether to allow access to members that are hidden by external modules.
+        skipFileNeededCheck?: boolean; // Whether to skip checking if the file is needed for the alias resolution.
+    }
+
+    // Describes a file that's been opened in the language server which the type server needs to be aware of.
+    export interface TextDocumentOpenParams {
+        uri: string;
+        text: string;
+        // The version of the file, which is used to track changes to the file. It will be incremented each time the file is modified.
+        version: number;
+        // Optional: If the file is part of a chain of files, this can be used to indicate the URI of the
+        // file that is automatically 'imported' by this file. This is useful for cases like Jupyter notebooks
+        chainedFileUri?: string;
+    }
+
+    // Describes a file that's been closed in the language server which the type server needs to be aware of.
+    export interface TextDocumentCloseParams {
+        uri: string;
+    }
+
+    // Parameters for resolving an import
+    export interface ResolveImportParams {
+        // The URI of the source file where the import is referenced.
+        sourceUri: string;
+        // The descriptor of the imported module.
+        moduleDescriptor: ImportedModuleDescriptor;
+        snapshot: number;
+    }
+
+    // Represents settings that can be sent to the type server.
+
+    export type Settings = {
+        [key: string]: any;
+    };
+
+    // Requests and notifications for the type server protocol.
+    export enum Requests {
+        // First request to initialize the type server.
+        Initialize = 'typeServer/initialize',
+        // Request from client to shut down the type server.
+        Shutdown = 'typeServer/shutdown',
+        // Request from client to get the current snapshot of the type server.
+        // A snapshot is a point-in-time representation of the type server's state, including all loaded files and their types.
+        // A type server should change its snapshot whenever any type it might have returned is no longer valid. Meaning types are
+        // only usable for the snapshot they were returned with.
+        GetSnapshot = 'typeServer/getSnapshot',
```

Comment:
I presume this returns some sort of an ID that uniquely identifies the current state? Or does it actively create a new snapshot object whose lifetime must be tracked and managed by the type server?

Reply:
ID that uniquely identifies the current state. 

The way I have it working right now with Pyright is that any invalidation of the cache causes the internal snapshot number to increment. So a snapshot is just a number. And it really just tracks if there's been any changes detected since the snapshot was handed out. 

It's not intended to be an actual snapshot of the state. If the snapshot doesn't match, all requests should return undefined for their results or throw an exception. Haven't worked that out yet. Exception is probably better. LSP already has this idea for canceled exceptions. We could add a new error result number that indicates invalid snapshot.

I'll add more comments to indicate that the server is not required to handle more than one snapshot at a time.